### PR TITLE
[Optimize Instructions] Combine reinterprets, loads and stores

### LINKED ISSUE #3973: Optimize load/store + reinterpret ops into one load / store operation
```wat
i64.load
f64.reinterpret_i64
```
into
```wat
f64.load
```

and 

```wat
i64.reinterpret_f64
i64.store
```
into
```wat
f64.store
```
the same for `f32.reinterpret_i32(i32.load(..))` -> `f32.load(..)` and `i32.store(.., f32.reinterpret_i32))` -> `f32.store(..)`

### PR DESCRIPTION
**Pull request by MaxGraey**
FIx #3973

```rust
f32.reinterpret_i32(i32.load(x))  =>  f32.load(x)
f64.reinterpret_i64(i64.load(x))  =>  f64.load(x)
i32.reinterpret_f32(f32.load(x))  =>  i32.load(x)
i64.reinterpret_f64(f64.load(x))  =>  i64.load(x)
```

```rust
f32.store(y, f32.reinterpret_i32(x))  =>  i32.store(y, x)
f64.store(y, f64.reinterpret_i64(x))  =>  i64.store(y, x)
i32.store(y, i32.reinterpret_f32(x))  =>  f32.store(y, x)
i64.store(y, i64.reinterpret_f64(x))  =>  f64.store(y, x)
```

```rust
i32.reinterpret_f32(f32.reinterpret_i32(x))  =>  x
i64.reinterpret_f64(f64.reinterpret_i64(x))  =>  x
f32.reinterpret_i32(i32.reinterpret_f32(x))  =>  x
f64.reinterpret_i64(i64.reinterpret_f64(x))  =>  x
```

### CODE REVIEW

**Comment by MaxGraey** [comment]
Fuzzed: `ITERATION: 21680`

**Comment by MaxGraey** [comment]
@tlively Could you take a look on this please? =)

File: src/ir/properties.h [hunk 1]
```diff
@@ -26,6 +26,24 @@ namespace wasm {
 
 namespace Properties {
 
+inline int getBits(const Type& type) {
```
Comment (kripken) [inline src/ir/properties.h:29]: Looks like https://github.com/WebAssembly/binaryen/blob/1e79c53ec1ad3d80db3354f15919331fdfa7ed28/src/wasm-type.h#L193-L194 ? (but in bits and not bytes)

File: test/lit/passes/optimize-instructions.wast [hunk 2]
```diff
@@ -12379,4 +12379,124 @@
...
+  ;; CHECK-NEXT:   (local.get $w)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $eliminate_reinterpret_reinterpret (param $x i32) (param $y i64) (param $z f32) (param $w f64)
+    (drop (i32.reinterpret_f32 (f32.reinterpret_i32 (local.get $x))))
+    (drop (i64.reinterpret_f64 (f64.reinterpret_i64 (local.get $y))))
+    (drop (f32.reinterpret_i32 (i32.reinterpret_f32 (local.get $z))))
+    (drop (f64.reinterpret_i64 (i64.reinterpret_f64 (local.get $w))))
+  )
 )
```
Comment (kripken) [inline test/lit/passes/optimize-instructions.wast:12502]: Please add testcases for skipping atomics. I think after that, all the conditions in the new code will be exercised.

File: src/ir/properties.h (hunk 1 above)
Comment (MaxGraey) [reply src/ir/properties.h:29]: Oh right! Totally forget about `getByteSize`!

**Comment by MaxGraey** [comment]
Done!

**Review by kripken — APPROVED** [review]
Great!

**Pull request merged**